goodware/putty/buildbogus/doc/psusan.1
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.TH "psusan" "1" "2020\(hy12\(hy13" "PuTTY\ tool\ suite" "PuTTY\ tool\ suite"
.SH "NAME"
.PP
\fBpsusan\fP \- pseudo-SSH for untappable, separately authenticated networks
.SH "SYNOPSIS"
.PP
.nf
\fBpsusan\fP\ [\ \fIoptions\fP\ ]
.fi
.SH "DESCRIPTION"
.PP
\fBpsusan\fP is a server program that behaves like the innermost `connection' layer of an SSH session, without the two outer security layers of encryption and authentication. It provides all the post-authentication features of an SSH connection:
.IP "\fB\(bu\fP"
choosing whether to run an interactive terminal session or a single specified command
.IP "\fB\(bu\fP"
multiple terminal sessions at once (or a mixture of those and specified commands)
.IP "\fB\(bu\fP"
SFTP file transfer
.IP "\fB\(bu\fP"
all the standard SSH port-forwarding options
.IP "\fB\(bu\fP"
X11 forwarding
.IP "\fB\(bu\fP"
SSH agent forwarding
.PP
The catch is that, because it lacks the outer layers of SSH, you have to run it over some kind of data channel that is already authenticated as the right user, and that is already protected to your satisfaction against eavesdropping and session hijacking. A good rule of thumb is that any channel that you were prepared to run a \fIbare\fP shell session over, you can run \fBpsusan\fP over instead, which adds all the above conveniences without changing the security properties.
.PP
The protocol that \fBpsusan\fP speaks is also spoken by PuTTY, Plink, PSCP, and PSFTP, if you select the protocol type `Bare ssh-connection' or the command-line option \fB-ssh-connection\fP and specify the absolute path to the appropriate Unix-domain socket in place of a hostname.
.SH "EXAMPLES"
.PP
The idea of a secure, pre-authenticated data channel seems strange to people thinking about \fInetwork\fP connections. But there are lots of examples within the context of a single Unix system, and that's where \fBpsusan\fP is typically useful.
.SS "Docker"
.PP
A good example is the console or standard I/O channel leading into a container or virtualisation system. Docker is a familiar example. If you want to start a Docker container and run a shell directly within it, you might say something like
.PP
.nf
docker\ run\ \-i\ \-t\ \fIsome:image\fP
.fi
.PP
which will allow you to run a single shell session inside the container, in the same terminal you started Docker from.
.PP
Suppose that you'd prefer to run \fImultiple\fP shell sessions in the same container at once (perhaps so that one of them can use debugging tools to poke at what another is doing). And perhaps inside that container you're going to run a program that you don't trust with full access to your network, but are prepared to let it make one or two specific network connections of the kind you could set up with an SSH port forwarding.
.PP
In that case, you could remove the \fB-t\fP option from that Docker command line (which means `allocate a terminal device'), and tell it to run \fBpsusan\fP inside the container:
.PP
.nf
docker\ run\ \-i\ \fIsome:image\fP\ /\fIsome/path/to\fP/psusan
.fi
.PP
(Of course, you'll need to ensure that \fBpsusan\fP is installed somewhere inside the container image.)
.PP
If you do that from a shell command line, you'll see a banner line looking something like this:
.PP
.nf
SSHCONNECTION@putty.projects.tartarus.org\-2.0\-PSUSAN_Release_0.75
.fi
.PP
which isn't particularly helpful except that it tells you that \fBpsusan\fP has started up successfully.
.PP
To talk to this server \fIusefully\fP, you can set up a PuTTY saved session as follows:
.IP "\fB\(bu\fP"
Set the protocol to `Bare ssh-connection' (the \fBpsusan\fP protocol).
.IP "\fB\(bu\fP"
Write \fIsomething\fP in the hostname box. It will appear in PuTTY's window title (if you run GUI PuTTY), so you might want to write something that will remind you what kind of window it is. If you have no opinion, something generic like `\fBdummy\fP' will do.
.IP "\fB\(bu\fP"
In the `Proxy' configuration panel, set the proxy type to `Local', and enter the above `\fBdocker run\fP' command in the `Telnet command, or local proxy command' edit box.
.IP "\fB\(bu\fP"
In the `SSH' configuration panel, you will very likely want to turn on connection sharing. (See below.)
.PP
This arranges that when PuTTY starts up, it will run the Docker command as shown above in place of making a network connection, and talk to that command using the \fBpsusan\fP SSH-like protocol.
.PP
The effect is that you will still get a shell session in the context of a Docker container. But this time, it's got all the SSH amenities. If you also turn on connection sharing in the `SSH' configuration panel, then the `Duplicate Session' option will get you a second shell in the \fIsame\fP Docker container (instead of a primary shell in a separate instance). You can transfer files in and out of the container while it's running using PSCP or PSFTP; you can forward network ports, X11 programs, and/or an SSH agent to the container.
.PP
Of course, another way to do all of this would be to run the \fIfull\fP SSH protocol over the same channel. This involves more setup: you have to invent an SSH host key for the container, accept it in the client, and deal with it being left behind in your client's host key cache when the container is discarded. And you have to set up some login details in the container: either configure a password, and type it in the client, or copy in the public half of some SSH key you already had. And all this inconvenience is \fIunnecessary\fP, because these are all precautions you need to take when the connection between two systems is going over a hostile network. In this case, it's only going over a kernel IPC channel that's guaranteed to go to the right place, so those safety precautions are redundant, and they only add awkwardness.
.SS "User-mode Linux"
.PP
User-mode Linux is another container type you can talk to in the same way. Here's a small worked example.
.PP
The \fIeasiest\fP way to run UML is to use its `\fBhostfs\fP' file system type to give the guest kernel access to the same virtual filesystem as you have on the host. For example, a command line like this gets you a shell prompt inside a UML instance sharing your existing filesystem:
.PP
.nf
linux\ mem=512M\ rootfstype=hostfs\ rootflags=/\ rw\ init=/bin/bash
.fi
.PP
If you run this at a command line (assuming you have a UML kernel available on your path under the name `\fBlinux\fP'), then you should see a lot of kernel startup messages, followed by a shell prompt along the lines of
.PP
.nf
root@(none):/#
.fi
.PP
To convert this into a \fBpsusan\fP-based UML session, we need to adjust the command line so that instead of running \fBbash\fP it runs \fBpsusan\fP. But running \fBpsusan\fP directly isn\*(Aqt quite enough, because \fBpsusan\fP will depend on a small amount of setup, such as having \fB/proc\fP mounted. So instead, we set the init process to a shell script which will do the necessary setup and \fIthen\fP invoke \fBpsusan\fP.
.PP
Also, running \fBpsusan\fP directly over the UML console device is a bad idea, because then the \fBpsusan\fP binary protocol will be mixed with textual console messages. So a better plan is to redirect UML\*(Aqs console to the standard error of the \fBlinux\fP process, and map its standard input and output to a serial port. So the replacement UML command line might look something like this:
.PP
.nf
linux\ mem=512M\ rootfstype=hostfs\ rootflags=/\ rw\ \e
\ \ \ \ con=fd:2,fd:2\ ssl0=fd:0,fd:1\ init=\fI/some/path/to/uml\-psusan.sh\fP
.fi
.PP
And the setup script \fBuml-psusan.sh\fP might look like this:
.PP
.nf
#!/bin/bash
\fI#\ Set\ up\ vital\ pseudo\-filesystems\fP
mount\ \-t\ proc\ none\ /proc
mount\ \-t\ devpts\ none\ /dev/pts
\fI#\ Redirect\ I/O\ to\ the\ serial\ port,\ but\ stderr\ to\ the\ console\fP
exec\ 0<>/dev/ttyS0\ 1>&0\ 2>/dev/console
\fI#\ Set\ the\ serial\ port\ into\ raw\ mode,\ to\ run\ a\ binary\ protocol\fP
stty\ raw\ \-echo
\fI#\ Choose\ what\ shell\ you\ want\ to\ run\ inside\ psusan\fP
export\ SHELL=/bin/bash
\fI#\ Set\ up\ a\ default\ path\fP
export\ PATH=/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/bin:/sbin
\fI#\ And\ now\ run\ psusan\ over\ the\ serial\ port\fP
exec\ /home/simon/src/putty/misc/psusan
.fi
.PP
Now set up a PuTTY saved session as in the Docker example above. Basically you'll want to use the above \fBlinux\fP command as the local proxy command. However, it\*(Aqs worth wrapping it in \fBsetsid\fP(\fI1\fP), because when UML terminates, it kills its entire process group. So it\*(Aqs better that PuTTY should not be part of that group, and should have the opportunity to shut down cleanly by itself. So probably you end up setting the proxy command to be something more like:
.PP
.nf
setsid\ linux\ mem=512M\ rootfstype=hostfs\ rootflags=/\ rw\ \e
\ \ \ \ con=fd:2,fd:2\ ssl0=fd:0,fd:1\ init=\fI/some/path/to/uml\-psusan.sh\fP
.fi
.PP
You may also find that you have to enable the bug workaround that indicates that the server `Discards data sent before its greeting', because otherwise PuTTY's outgoing protocol greeting can be accidentally lost during UML startup. (See Debian bug #991958.)
.PP
Once you've done that, you'll have a PuTTY session that starts up a clean UML instance when you run it, and (if you enabled connection sharing) further instances of the same session will connect to the same instance again.
.SS "Windows Subsystem for Linux"
.PP
On Windows, the default way to use WSL is to run the \fBwsl\fP program, or one of its aliases, in a Windows console, either by launching it from an existing command prompt, or by using a shortcut that opens it in a fresh console. This gives you a Linux terminal environment, but in a Windows console window.
.PP
If you'd prefer to interact with the same environment using PuTTY as the terminal (for example, if you prefer PuTTY's mouse shortcuts for copy and paste), you can set it up by installing \fBpsusan\fP in the Linux environment, and then setting up a PuTTY saved session that talks to it. A nice way to do this is to use the name of the WSL distribution as the `host name':
.IP "\fB\(bu\fP"
set the local proxy command to `\fBwsl -d %host /usr/local/bin/psusan\fP' (or wherever you installed \fBpsusan\fP in the Linux system)
.IP "\fB\(bu\fP"
enter the name of a particular WSL distribution in the host name box. (For example, if you installed WSL Debian in the standard way from the Windows store, this will just be `Debian'.)
.IP "\fB\(bu\fP"
set the protocol to `Bare ssh-connection', as usual.
.PP
Like all the other examples here, this also permits you to forward ports in and out of the WSL environment (e.g. expose a WSL2 network service through the hypervisor's internal NAT), forward Pageant into it, and so on.
.SS "Cygwin"
.PP
Another Unix-like environment on Windows is Cygwin. That comes with its own GUI terminal application, \fBmintty\fP (as it happens, a derivative of PuTTY); but if you\*(Aqd prefer to use PuTTY itself to talk to your Cygwin terminal sessions, \fBpsusan\fP can help.
.PP
To do this, you'll first need to build the Unix PuTTY tools inside Cygwin (via the usual \fBcmake\fP method). Then, copy the resulting \fBpsusan.exe\fP into Cygwin\*(Aqs \fB/bin\fP directory. (It has to be in that directory for non-Cygwin programs to run it; otherwise it won\*(Aqt be able to find the Cygwin DLL at startup.)
.PP
Then set up your PuTTY saved session like this:
.IP "\fB\(bu\fP"
set the local proxy command to run \fBpsusan.exe\fP via its real Windows path. You might also want to add the \fB--sessiondir\fP option so that shell sessions start up in your Cygwin home directory. For example, you might use the command `\fBc:\ecygwin64\ebin\epsusan.exe --sessiondir /home/simon\fP' (changing the pathname and username to match your setup).
.IP "\fB\(bu\fP"
enter anything you like in the host name box; `\fBCygwin\fP' is probably a good choice
.IP "\fB\(bu\fP"
set the protocol to `Bare ssh-connection', as usual.
.PP
Port forwarding is probably not particularly useful in this case, since Cygwin shares the same network port space as the host machine. But turning on agent forwarding is useful, because then the Cygwin command-line SSH client can talk to Pageant without any further configuration.
.SS "\fBschroot\fP"
.PP
Another example of a container-like environment is the alternative filesystem layout set up by \fBschroot\fP(\fI1\fP).
.PP
\fBschroot\fP is another program that defaults to running an interactive shell session in the terminal you launched it from. But again, you can get a \fBpsusan\fP connection into the \fBschroot\fP environment by setting up a PuTTY saved session whose local proxy command is along the lines of
.PP
.nf
schroot\ \-c\ \fIchroot\-name\fP\ /\fIsome/path/to\fP/psusan
.fi
.PP
Depending on how much of the chroot environment is copied from your main one, you might find this makes it easier to (for example) run X11 programs inside the chroot that open windows on your main X display, or transfer files in and out of the chroot.
.SS "Between network namespaces"
.PP
If you've set up multiple network namespaces on a Linux system, with different TCP/IP configurations, then \fBpsusan\fP can be a convenient unprivileged-user gateway between them, if you run it as a non-root user in the non-default one of your namespaces, listening for connections on a Unix-domain socket.
.PP
If you do that, then it gives you convenient control over which of your outgoing network connections use which TCP/IP configuration: you can use PuTTY to run a shell session in the context of the other namespace if you want to run commands like \fBping\fP, or you can set up individual port forwardings or even a SOCKS server so that processes running in one namespace can send their network connections via the other one.
.PP
For this application, it's probably most convenient to use the \fB--listen\fP option in \fBpsusan\fP, which makes it run as a server and listen for connections on a Unix-domain socket. Then you can enter that socket name in PuTTY\*(Aqs host name configuration field (and also still select the `Bare ssh-connection' protocol option), to connect to that socket as if it were an SSH client.
.PP
Provided the Unix-domain socket is inside a directory that only the right user has access to, this will ensure that authentication is done implicitly by the Linux kernel.
.SS "Between user ids, via GNU userv"
.PP
If you use multiple user ids on the same machine, say for purposes of privilege separation (running some less-trusted program with limited abilities to access all your stuff), then you probably have a `default' or most privileged account where you run your main login session, and sometimes need to run a shell in another account.
.PP
\fBpsusan\fP can be used as an access channel between the accounts, using GNU \fBuserv\fP(\fI1\fP) as the transport. In the account you want to access, write a \fBuserv\fP configuration stanza along the lines of
.PP
.nf
if\ (glob\ service\ psusan\ &\ glob\ calling\-user\ \fImy\-main\-account\-name\fP)
\ \ \ reset
\ \ \ execute\ /\fIsome/path/to\fP/psusan
fi
.fi
.PP
This gives your main account the right to run the command
.PP
.nf
userv\ \fImy\-sub\-account\-name\fP\ psusan
.fi
.PP
and you can configure that command name as a PuTTY local proxy command, in the same way as most of the previous examples.
.PP
Of course, there are plenty of ways already to access one local account from another, such as \fBsudo\fP. One advantage of doing it this way is that you don\*(Aqt need the system administrator to intervene when you want to change the access controls (e.g. change which of your accounts have access to another): as long as you have \fIsome\fP means of getting into each account in the first place, and \fBuserv\fP is installed, you can make further configuration changes without having to bother root about it.
.PP
Another advantage is that it might make file transfer between the accounts easier. If you're the kind of person who keeps your home directories private, then it's awkward to copy a file from one of your accounts to another just by using the \fBcp\fP command, because there\*(Aqs nowhere convenient that you can leave it in one account where the other one can read it. But with \fBpsusan\fP over \fBuserv\fP, you don\*(Aqt need any shared piece of filesystem: you can \fBscp\fP files back and forth without any difficulty.
.SH "OPTIONS"
.PP
The command-line options supported by \fBpsusan\fP are:
.IP "\fB--listen\fP \fIunix-socket-name\fP"
Run \fBpsusan\fP in listening mode. \fIunix-socket-name\fP is the pathname of a Unix-domain socket to listen on. You should ensure that this pathname is inside a directory whose read and exec permissions are restricted to only the user(s) you want to be able to access the environment that \fBpsusan\fP is running in.
.RS
.PP
The listening socket has to be a Unix-domain socket. \fBpsusan\fP does not provide an option to run over TCP/IP, because the unauthenticated nature of the protocol would make it inherently insecure.
.RE
.IP "\fB--listen-once\fP"
In listening mode, this option causes \fBpsusan\fP to listen for only one connection, and exit immediately after that connection terminates.
.IP "\fB--sessiondir\fP \fIpathname\fP"
This option sets the directory that shell sessions and subprocesses will start in. By default it is \fBpsusan\fP\*(Aqs own working directory, but in some situations it\*(Aqs easier to change it with a command-line option than by wrapping \fBpsusan\fP in a script that changes directory before starting it.
.IP "\fB-v\fP, \fB--verbose\fP"
This option causes \fBpsusan\fP to print verbose log messages on its standard error. This is probably most useful in listening mode.
.IP "\fB\-sshlog\fP \fIlogfile\fP"

.IP "\fB\-sshrawlog\fP \fIlogfile\fP"
These options cause \fBpsusan\fP to log protocol details to a file, similarly to the logging options in PuTTY and Plink.
.RS
.PP
\fB\-sshlog\fP logs decoded SSH packets and other events (those that \fB\-v\fP would print). \fB\-sshrawlog\fP additionally logs the raw wire data, including the outer packet format and the initial greetings. 
.RE


goodware/putty/unix/psusan.c
/*
 * 'psusan': Pseudo Ssh for Untappable, Separately Authenticated Networks
 *
 * This is a standalone application that speaks on its standard I/O
 * (or a listening Unix-domain socket) the server end of the bare
 * ssh-connection protocol used by PuTTY's connection sharing.
 *
 * The idea of this tool is that you can use it to communicate across
 * any 8-bit-clean data channel between two inconveniently separated
 * domains, provided the channel is already (as the name suggests)
 * adequately secured against eavesdropping and modification and
 * already authenticated as the right user.
 *
 * If you're sitting at one end of such a channel and want to type
 * commands into the other end, the most obvious thing to do is to run
 * a terminal session directly over it. But if you run psusan at one
 * end, and a PuTTY (or compatible) client at the other end, then you
 * not only get a single terminal session: you get all the other SSH
 * amenities, like the ability to spawn extra terminal sessions,
 * forward ports or X11 connections, even forward an SSH agent.
 *
 * There are a surprising number of channels of that kind; see the man
 * page for some examples.
 */

#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <assert.h>
#include <stdarg.h>
#include <signal.h>
#include <unistd.h>
#include <fcntl.h>
#include <termios.h>
#include <pwd.h>
#include <sys/ioctl.h>
#include <sys/time.h>

#include "putty.h"
#include "mpint.h"
#include "ssh.h"
#include "ssh/server.h"

void modalfatalbox(const char *p, ...)
{
    va_list ap;
    fprintf(stderr, "FATAL ERROR: ");
    va_start(ap, p);
    vfprintf(stderr, p, ap);
    va_end(ap);
    fputc('\n', stderr);
    exit(1);
}
void nonfatal(const char *p, ...)
{
    va_list ap;
    fprintf(stderr, "ERROR: ");
    va_start(ap, p);
    vfprintf(stderr, p, ap);
    va_end(ap);
    fputc('\n', stderr);
}

char *platform_default_s(const char *name)
{
    return NULL;
}

bool platform_default_b(const char *name, bool def)
{
    return def;
}

int platform_default_i(const char *name, int def)
{
    return def;
}

FontSpec *platform_default_fontspec(const char *name)
{
    return fontspec_new("");
}

Filename *platform_default_filename(const char *name)
{
    return filename_from_str("");
}

char *x_get_default(const char *key)
{
    return NULL;                       /* this is a stub */
}

void old_keyfile_warning(void) { }

void timer_change_notify(unsigned long next)
{
}

char *platform_get_x_display(void) { return NULL; }

void make_unix_sftp_filehandle_key(void *vdata, size_t size)
{
    /* psusan runs without a random number generator, so we can't make
     * this up by random_read. Fortunately, psusan is also
     * non-adversarial, so it's safe to generate this trivially. */
    unsigned char *data = (unsigned char *)vdata;
    for (size_t i = 0; i < size; i++)
        data[i] = (unsigned)rand() / ((unsigned)RAND_MAX / 256);
}

static bool verbose;

struct server_instance {
    unsigned id;
    LogPolicy logpolicy;
};

static void log_to_stderr(unsigned id, const char *msg)
{
    if (!verbose)
        return;
    if (id != (unsigned)-1)
        fprintf(stderr, "#%u: ", id);
    fputs(msg, stderr);
    fputc('\n', stderr);
    fflush(stderr);
}

static void server_eventlog(LogPolicy *lp, const char *event)
{
    struct server_instance *inst = container_of(
        lp, struct server_instance, logpolicy);
    if (verbose)
        log_to_stderr(inst->id, event);
}

static void server_logging_error(LogPolicy *lp, const char *event)
{
    struct server_instance *inst = container_of(
        lp, struct server_instance, logpolicy);
    log_to_stderr(inst->id, event);    /* unconditional */
}

static int server_askappend(
    LogPolicy *lp, Filename *filename,
    void (*callback)(void *ctx, int result), void *ctx)
{
    return 2; /* always overwrite (FIXME: could make this a cmdline option) */
}

static const LogPolicyVtable server_logpolicy_vt = {
    .eventlog = server_eventlog,
    .askappend = server_askappend,
    .logging_error = server_logging_error,
    .verbose = null_lp_verbose_no,
};

static void show_help(FILE *fp)
{
    fputs("usage:   psusan [options]\n"
          "options: --listen SOCKETPATH  listen for connections on a Unix-domain socket\n"
          "         --listen-once        (with --listen) stop after one connection\n"
          "         --verbose            print log messages to standard error\n"
          "         --sessiondir DIR     cwd for session subprocess (default $HOME)\n"
          "         --sshlog FILE        write ssh-connection packet log to FILE\n"
          "         --sshrawlog FILE     write packets and raw data log to FILE\n"
          "also:    psusan --help        show this text\n"
          "         psusan --version     show version information\n", fp);
}

static void show_version_and_exit(void)
{
    char *buildinfo_text = buildinfo("\n");
    printf("%s: %s\n%s\n", appname, ver, buildinfo_text);
    sfree(buildinfo_text);
    exit(0);
}

const bool buildinfo_gtk_relevant = false;

static bool listening = false, listen_once = false;
static bool finished = false;
void server_instance_terminated(LogPolicy *lp)
{
    struct server_instance *inst = container_of(
        lp, struct server_instance, logpolicy);

    if (listening && !listen_once) {
        log_to_stderr(inst->id, "connection terminated");
    } else {
        finished = true;
    }

    sfree(inst);
}

bool psusan_continue(void *ctx, bool fd, bool cb)
{
    return !finished;
}

static bool longoptarg(const char *arg, const char *expected,
                       const char **val, int *argcp, char ***argvp)
{
    int len = strlen(expected);
    if (memcmp(arg, expected, len))
        return false;
    if (arg[len] == '=') {
        *val = arg + len + 1;
        return true;
    } else if (arg[len] == '\0') {
        if (--*argcp > 0) {
            *val = *++*argvp;
            return true;
        } else {
            fprintf(stderr, "%s: option %s expects an argument\n",
                    appname, expected);
            exit(1);
        }
    }
    return false;
}

static bool longoptnoarg(const char *arg, const char *expected)
{
    int len = strlen(expected);
    if (memcmp(arg, expected, len))
        return false;
    if (arg[len] == '=') {
        fprintf(stderr, "%s: option %s expects no argument\n",
                appname, expected);
        exit(1);
    } else if (arg[len] == '\0') {
        return true;
    }
    return false;
}

struct server_config {
    Conf *conf;
    const SshServerConfig *ssc;

    unsigned next_id;

    Socket *listening_socket;
    Plug listening_plug;
};

static Plug *server_conn_plug(
    struct server_config *cfg, struct server_instance **inst_out)
{
    struct server_instance *inst = snew(struct server_instance);

    memset(inst, 0, sizeof(*inst));

    inst->id = cfg->next_id++;
    inst->logpolicy.vt = &server_logpolicy_vt;

    if (inst_out)
        *inst_out = inst;

    return ssh_server_plug(
        cfg->conf, cfg->ssc, NULL, 0, NULL, NULL,
        &inst->logpolicy, &unix_live_sftpserver_vt);
}

static void server_log(Plug *plug, PlugLogType type, SockAddr *addr, int port,
                       const char *error_msg, int error_code)
{
    log_to_stderr(-1, error_msg);
}

static void server_closing(Plug *plug, PlugCloseType type,
                           const char *error_msg)
{
    if (type != PLUGCLOSE_NORMAL)
        log_to_stderr(-1, error_msg);
}

static int server_accepting(Plug *p, accept_fn_t constructor, accept_ctx_t ctx)
{
    struct server_config *cfg = container_of(
        p, struct server_config, listening_plug);
    Socket *s;
    const char *err;

    struct server_instance *inst;

    if (listen_once) {
        if (!cfg->listening_socket) /* in case of rapid double-accept */
            return 1;
        sk_close(cfg->listening_socket);
        cfg->listening_socket = NULL;
    }

    Plug *plug = server_conn_plug(cfg, &inst);
    s = constructor(ctx, plug);
    if ((err = sk_socket_error(s)) != NULL)
        return 1;

    SocketPeerInfo *pi = sk_peer_info(s);

    char *msg = dupprintf("new connection from %s", pi->log_text);
    log_to_stderr(inst->id, msg);
    sfree(msg);
    sk_free_peer_info(pi);

    sk_set_frozen(s, false);
    ssh_server_start(plug, s);
    return 0;
}

static const PlugVtable server_plugvt = {
    .log = server_log,
    .closing = server_closing,
    .accepting = server_accepting,
};

unsigned auth_methods(AuthPolicy *ap)
{ return 0; }
bool auth_none(AuthPolicy *ap, ptrlen username)
{ return false; }
int auth_password(AuthPolicy *ap, ptrlen username, ptrlen password,
                  ptrlen *new_password_opt)
{ return 0; }
bool auth_publickey(AuthPolicy *ap, ptrlen username, ptrlen public_blob)
{ return false; }
RSAKey *auth_publickey_ssh1(
    AuthPolicy *ap, ptrlen username, mp_int *rsa_modulus)
{ return NULL; }
AuthKbdInt *auth_kbdint_prompts(AuthPolicy *ap, ptrlen username)
{ return NULL; }
int auth_kbdint_responses(AuthPolicy *ap, const ptrlen *responses)
{ return -1; }
char *auth_ssh1int_challenge(AuthPolicy *ap, unsigned method, ptrlen username)
{ return NULL; }
bool auth_ssh1int_response(AuthPolicy *ap, ptrlen response)
{ return false; }
bool auth_successful(AuthPolicy *ap, ptrlen username, unsigned method)
{ return false; }

int main(int argc, char **argv)
{
    const char *listen_socket = NULL;

    SshServerConfig ssc;

    Conf *conf = make_ssh_server_conf();

    memset(&ssc, 0, sizeof(ssc));

    ssc.application_name = "PSUSAN";
    ssc.session_starting_dir = getenv("HOME");
    ssc.bare_connection = true;

    while (--argc > 0) {
        const char *arg = *++argv;
        const char *val;

        if (longoptnoarg(arg, "--help")) {
            show_help(stdout);
            exit(0);
        } else if (longoptnoarg(arg, "--version")) {
            show_version_and_exit();
        } else if (longoptnoarg(arg, "--verbose") || !strcmp(arg, "-v")) {
            verbose = true;
        } else if (longoptarg(arg, "--sessiondir", &val, &argc, &argv)) {
            ssc.session_starting_dir = val;
        } else if (longoptarg(arg, "--sshlog", &val, &argc, &argv) ||
                   longoptarg(arg, "-sshlog", &val, &argc, &argv)) {
            Filename *logfile = filename_from_str(val);
            conf_set_filename(conf, CONF_logfilename, logfile);
            filename_free(logfile);
            conf_set_int(conf, CONF_logtype, LGTYP_PACKETS);
            conf_set_int(conf, CONF_logxfovr, LGXF_OVR);
        } else if (longoptarg(arg, "--sshrawlog", &val, &argc, &argv) ||
                   longoptarg(arg, "-sshrawlog", &val, &argc, &argv)) {
            Filename *logfile = filename_from_str(val);
            conf_set_filename(conf, CONF_logfilename, logfile);
            filename_free(logfile);
            conf_set_int(conf, CONF_logtype, LGTYP_SSHRAW);
            conf_set_int(conf, CONF_logxfovr, LGXF_OVR);
        } else if (longoptarg(arg, "--listen", &val, &argc, &argv)) {
            listen_socket = val;
        } else if (!strcmp(arg, "--listen-once")) {
            listen_once = true;
        } else {
            fprintf(stderr, "%s: unrecognised option '%s'\n", appname, arg);
            exit(1);
        }
    }

    sk_init();
    uxsel_init();

    struct server_config scfg;
    scfg.conf = conf;
    scfg.ssc = &ssc;
    scfg.next_id = 0;

    if (listen_socket) {
        listening = true;
        scfg.listening_plug.vt = &server_plugvt;
        SockAddr *addr = unix_sock_addr(listen_socket);
        scfg.listening_socket = new_unix_listener(addr, &scfg.listening_plug);
        char *msg = dupprintf("listening on Unix socket %s", listen_socket);
        log_to_stderr(-1, msg);
        sfree(msg);
    } else {
        struct server_instance *inst;
        Plug *plug = server_conn_plug(&scfg, &inst);
        ssh_server_start(plug, make_fd_socket(0, 1, -1, NULL, 0, plug));
        log_to_stderr(inst->id, "running directly on stdio");
    }

    cli_main_loop(cliloop_no_pw_setup, cliloop_no_pw_check,
                  psusan_continue, NULL);

    return 0;
}
